### PR TITLE
Deploy Metabase and demo instance with access to it

### PR DESCRIPTION
**Pull request by AndrewDryga**

### CODE REVIEW

File: terraform/modules/gateway-google-cloud-compute/main.tf [hunk 1]
```diff
@@ -46,6 +46,20 @@ locals {
     {
       name  = "FIREZONE_API_URL"
       value = var.api_url
+    },
+    # TODO: this is really confusing to set here, having explicit ID
+    # doesn't play well with horizontal replicas
+    {
```
Comment (AndrewDryga) [inline terraform/modules/gateway-google-cloud-compute/main.tf:52]: @jamilbk I was feeling confused here for some time thinking what to do with this ID, I bet clients trying to replicate this horizontally will have a hard time too

File: terraform/environments/staging/demo.tf [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+       | sudo gpg --dearmor -o /usr/share/keyrings/docker-archive-keyring.gpg \
+    && echo "deb [arch=amd64 signed-by=/usr/share/keyrings/docker-archive-keyring.gpg] https://download.docker.com/linux/ubuntu $(lsb_release -cs) stable" \
+       | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null \
+    && sudo apt update -y \
+    && sudo apt install docker-ce docker-ce-cli containerd.io -y \
+    && sudo usermod -aG docker $(whoami) \
+    && sudo systemctl enable docker \
+    && sudo systemctl start docker \
+    && sudo docker run -d --restart always --name=httpbin -p 80:80 kennethreitz/httpbin \
+    && echo ${module.metabase.internal_ip} metabase.fz >> /etc/hosts
```
Comment (AndrewDryga) [inline terraform/environments/staging/demo.tf:69]: This allows us to create a:
- DNS resource for `metabase.fz`
- CIDR resource for apps subnet `10.128.0.0/20`
- IP resource for Postgres database (see `terraform outputs` to see it's internal IP)
- localhost iperf3 resource `127.0.0.1/32`
- locahost docker container resource (HTTPBin) `localhost:80`

File: terraform/modules/gateway-google-cloud-compute/main.tf (hunk 1 above)
Comment (bmanifold) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: Not that it has to change in this PR, but couldn't you just use the ID as the "base" and then append an integer at the point of use by using a terraform count index?

File: elixir/apps/web/lib/web/live/sites/show.ex [hunk 3]
```diff
@@ -59,71 +59,85 @@ defmodule Web.Sites.Show do
...
+              <div class="flex justify-center text-center text-slate-500 p-4">
+                <div class="w-auto">
+                  <div class="pb-4">
+                    No gateway instances to display.
+                  </div>
+                  <div class="pb-4">
+                    TODO: Gateways should be deployed to your infrastructure,
+                    all gateways within a site should be able to access all site resources,
+                    connections within a site will be load balanced if there are multiple gateways deployed.
+                  </div>
```
Comment (bmanifold) [inline elixir/apps/web/lib/web/live/sites/show.ex]: How would you feel about holding off on this message until we have exactly what we want to say here?

**Review by bmanifold — COMMENTED** [review]
Overall, looks good, just one small question

File: terraform/modules/gateway-google-cloud-compute/main.tf (hunk 1 above)
Comment (AndrewDryga) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: @bmanifold there is no count for the instance template - we create just one, it doesn't matter how many actual instances are deployed from it. The same will happen with the AWS Instance Template and K8S Deployment yaml file.

So this decision to require `FIREZONE_ID` will force anyone doing it "the right way" writing custom environment-specific scripts trying to come up with a good way to uniquely identify an instance. (I'm not even sure that will be possible in all the environments, eg. if you use `gke-custom-metadata` aka a simple way to deploy containers to Google Compute - it won't be possible.)

I've used `hostname` as a hack but this is not the best idea, and it is pretty complicated if you think about it as a DevOps. Overall we will have to write 3-10x the amount of documentation and deployment examples just to keep it, here is a short intro:

## Kubernetes

For k8s the best way will be something like this:
```
- name: FIREZONE_ID
   valueFrom:
     fieldRef:
       fieldPath: spec.nodeName
```
pod names are unique so this might suffice, but on any deployment (or node upgrades) new pods will be created so new gateways will be created anyway, and this doesn't solve the issue of having a large amount of gateways. (So no charging per gateway is possible.)

The solution might be to enforce people using Stateful Sets for k8s (that have pods with static indexed names) but it only works if you want to have a bastion-like deployment. 

We can also use Daemon Sets and read actual node name but this will deploy a gateway to every node in a cluster, which can or can not be desirable either.

To sum up: gateways will be created with each pod (eg. sidecar deployment or just using regular deployment to create a replica set for our gateways) or we must enforce very strict deployment practices.

## Google Cloud

We will need to use a curl request and make sure that the instance service account token can access the metadata endpoint:
```
FIREZONE_ID=$(curl "http://metadata.google.internal/computeMetadata/v1/instance/id" -H "Metadata-Flavor: Google" -s)
# this will be set to 5895196805685327019
```
or
```
curl "http://metadata.google.internal/computeMetadata/v1/instance/hostname" -H "Metadata-Flavor: Google" -s
# this will be set to api-mftf.us-east1-d.c.firezone-staging.internal so I think we need to also shasum it to make sure invalid characters or too long strings won't be used for ID
```
but both will create new gateways on every deployment action. I don't know what is a good way to reliably solve it here other than not using Instance Templates / Groups and deploy barebone VMs that are harder to manage.

## AWS

Same as Google Cloud:
```
TOKEN=`curl -X PUT "http://169.254.169.254/latest/api/token" -q -H "X-aws-ec2-metadata-token-ttl-seconds: 21600"` \
&& curl -H "X-aws-ec2-metadata-token: $TOKEN" http://169.254.169.254/latest/meta-data/mac
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100    56  100    56    0     0  29787      0 --:--:-- --:--:-- --:--:-- 56000
12:ce:38:0b:f8:a1
```

## Bare metal or static VMs

Here everything is easier, you set the ID explicitly on every machine and make sure it doesn't change during reboots or maintenance operations..

File: elixir/apps/web/lib/web/live/sites/show.ex (hunk 3 above)
Comment (AndrewDryga) [reply elixir/apps/web/lib/web/live/sites/show.ex]: Was thinking of using some GPT magic to fill it before this PR is merged :).

File: terraform/modules/gateway-google-cloud-compute/main.tf (hunk 1 above)
Comment (bmanifold) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: @AndrewDryga, yeah, totally understand that if you wanted to use AWS Auto-scaling groups (or GCP equivalent) my suggestion of the Terraform count index wouldn't work.

That said, have we had any customers asking/wanting to use the deployment methods you've described?  This is probably a question and topic better discussed in real-time rather rather than async.

Comment (AndrewDryga) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: Yes, let's discuss it on a standup again.

I believe that GCP Instance Group is the recommended way to manage instances, no matter how many you want to deploy. Because it's the only way to get zero-downtime deployments, health checks, cross-az availability and automatic recovery. I think on AWS situation will be similar. It's also an easier way to manage a set of gateways - if you have more than one why would you use individual instances at all?..

Comment (bmanifold) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: > Because it's the only way to get zero-downtime deployments, health checks, cross-az availability and automatic recovery

I think I'd disagree that it's the only way to accomplish these things.  But, as we've said, let's discuss in real-time

Comment (AndrewDryga) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: cc @jamilbk every I try need to deploy firezone somewhere not as "one-off container" trying to prepare for the demo this is a huge pain in the ass - trying to put a bash script into a container environment variable.

Comment (jamilbk) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: Hm, maybe I'm missing something, but why can't the customer just statically define their gateways' `FIREZONE_ID`'s like they define other terraform resources?

They'd be required to use a unique ID for each gateway, but I don't think that causes a problem for the foreseeable future based on expected use cases of having 3x gateways per site. It'd probably map to how they'll be thinking of gateways initially as infrastructure elements.

In other words, I'm not seeing why the ID needs to be dynamically generated, hence the motivation behind just using `TOKEN` to identify the gateway.

Comment (AndrewDryga) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: Because it is impossible to do for pretty much anything (see examples above) except if you provision VMs one by one. And there is no good way to auto-generate it on all the platforms either - all of them will force us to think about gateways as ephemeral entities (as originally designed) rather than static.

Most idiomatic way to deploy anything these days are Instance Groups, K8S deployments, etc and they all are enforcing same environment variables for all the replicas.

Managing instances one by one is pain the ass if you need any advanced features (eg. zero downtime upgrades on pretty much any cloud platform) or even most simple things as liveness probes on Google Cloud.

Comment (AndrewDryga) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: And we should forget about running in a sidecar.

Comment (bmanifold) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: Can we put this on the agenda to discuss at standup tomorrow?  I don't think we actually ever talked this through.

Comment (jamilbk) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: @bmanifold sgtm

Comment (jamilbk) [reply terraform/modules/gateway-google-cloud-compute/main.tf:52]: Went back and re-read the thread, gave it more thought, and I have a better idea of the concern @AndrewDryga has. Will be prepared to discuss at standup
